### PR TITLE
Add Enumerable#find_by

### PR DESCRIPTION
**Pull request by americodls**
Add `Enumerable#find_by` that mimic  `ActiveRecord::FinderMethods#find_by`, as other methods do, like [`minimum`](https://github.com/rails/rails/blob/25396b1/activesupport/lib/active_support/core_ext/enumerable.rb#L40-L46), [`maximum`](https://github.com/rails/rails/blob/25396b1/activesupport/lib/active_support/core_ext/enumerable.rb#L48-L54), [`pluck`](https://github.com/rails/rails/blob/25396b1/activesupport/lib/active_support/core_ext/enumerable.rb#L189-L203), [`pick`](https://github.com/rails/rails/blob/25396b1/activesupport/lib/active_support/core_ext/enumerable.rb#L205-L220) and others.

### Motivation / Background

Would be nice if instead of this

```ruby
people.find do |person|
  person.name == 'John' && person.age == 42
end
```

We could do this

```ruby
people.find_by(name: 'John', age: 42)
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] There are no typos in commit messages and comments.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Feature branch is up-to-date with `main` (if not - rebase it).
* [x] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive.
* [x] Tests are added if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] PR is not in a draft state.
* [x] CI is passing.

### CODE REVIEW

**Comment by fatkodima** [comment]
Nice! You could also refactor code in several places, to demonstrate how this will improve the readability. For example, https://github.com/rails/rails/blob/ee81a0ed3407aa9cae68733c2380a9a49e088deb/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L968 https://github.com/rails/rails/blob/ee81a0ed3407aa9cae68733c2380a9a49e088deb/activerecord/lib/active_record/migration.rb#L982 and many more.

**Comment by americodls** [comment]
@fatkodima actually `find_by` would be a replacement for `find { |e| e.name == name }`, but not for `detect` or `select`, since it returns only the first occurrence.

**Comment by americodls** [comment]
@fatkodima I just added a new commit with the refactorings you suggested.

Probably not all places were refactored (I created Custom Cop to help me find and fix the simplest cases).

Let me know if I should improve something more here. 👋🏼

**Comment by skipkayhil** [comment]
In the related methods you linked there are two different implementations for what is being checked:
- for `maximum`/`minimum`, its an attribute (like `Payment.price`)
- for `pluck`/`pick`, its a hash value (like `element[:price]`)

It looks like this is currently doing the same as `maximum`/`minimum`, but I could see someone expecting the `pluck`/`pick` version. I don't think that a generic class like Enumerable should be dictating/knowing the shape of the objects inside it, so I'm not sure it's a good idea to include this.

File: activesupport/lib/active_support/core_ext/enumerable.rb [hunk 1]
```diff
@@ -110,6 +110,30 @@ def index_by
...
+  #
+  #   people.find_by(first_name: 'John', last_name: 'Doe')
+  #   # => #<Person first_name: "John", last_name: "Doe", age: 42>
+  #
+  #   people.find_by(last_name: 'Doe')
+  #   # => #<Person first_name: "Jane", last_name: "Doe", age: 23>
+  #
+  #   people.find_by(age: 32)
+  #   # => #<Person first_name: "John", last_name: "Doe", age: 32>
+  def find_by(attrs)
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/core_ext/enumerable.rb:129]: We need to document in the core extensions guide as well. https://github.com/rails/rails/blob/main/guides/source/active_support_core_extensions.md

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: activesupport/lib/active_support/core_ext/enumerable.rb (hunk 1 above)
Comment (americodls) [reply activesupport/lib/active_support/core_ext/enumerable.rb:129]: Added. Let me know if something else is needed. 😃

File: activesupport/lib/active_support/core_ext/enumerable.rb [hunk 2]
```diff
@@ -110,6 +110,30 @@ def index_by
     end
   end
 
+  # Finds the first element matching the specified attributes.
```
Comment (MatheusRich) [inline activesupport/lib/active_support/core_ext/enumerable.rb:113]: How about making it a bit more explicit that *all* attributes should match?

```suggestion
  # Finds the first element matching all the specified attributes.
```

**Comment by rafaelfranca** [comment]
After further consideration the core team decided to not add this method.

There is the issue of inconsistency between how to access the values (`#[]` vs `send`) but there is also the inconsistency on how Active Record's version work. An implementation on Enumerable would allow filtering on full methods, while Active Record's version only work with database columns.

For example if you had a `full_name` method in the object implemented as `first_name + last_name`, `find_by full_name: "Rafael França"` would work in a collection of loaded records, but not in a relation. This would cause a lot, and we don't think the usefulness of this method warrants all that possible confusion.

Thank you for the pull request.

**Pull request closed without merging**